src/main/scala/ai/eto/rikai/sql/spark/parser/SchemaParser.scala

```diff
@@ -0,0 +1,116 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ai.eto.rikai.sql.spark.parser
+
+import ai.eto.rikai.sql.spark.parser.RikaiModelSchemaParser.{
+  ArrayTypeContext,
+  PlainFieldTypeContext,
+  StructFieldContext,
+  StructTypeContext
+}
+import org.antlr.v4.runtime.atn.PredictionMode
+import org.antlr.v4.runtime.{CharStreams, CommonTokenStream}
+import org.apache.spark.sql.rikai.UDTCollection
+import org.apache.spark.sql.types._
+
+import scala.collection.JavaConverters.iterableAsScalaIterableConverter
+
+/**
+  * Build Spark's DataType from the model schema AST.
+  *
+  * It accepts any forms of value from `DataType.simpleString()`:
+  *
+  * {{{
+  *   * struct<score:int, box:box2d, label:string>
+  *   * struct<id:int, detection:array<struct<box:box2d, class:int>>>
+  *   * array<float>
+  *   * Primitive type: float
+  * }}}
+  */
+private class SchemaBuilder extends RikaiModelSchemaBaseVisitor[AnyRef] {
+
+  override def visitStructType(ctx: StructTypeContext): StructType = {
+    new StructType(
+      ctx.field.asScala.map(visit).map(v => v.asInstanceOf[StructField]).toArray
+    )
+  }
+
+  override def visitStructField(ctx: StructFieldContext): StructField = {
+    StructField(
+      ctx.name.getText,
+      visit(ctx.fieldType).asInstanceOf[DataType]
+    )
+  }
+
+  override def visitArrayType(ctx: ArrayTypeContext): ArrayType = {
+    ArrayType(visit(ctx.fieldType()).asInstanceOf[DataType])
+  }
+
+  override def visitPlainFieldType(ctx: PlainFieldTypeContext): DataType = {
+    val typeName = ctx.identifier().getText
+    typeName.toLowerCase match {
+      case "boolean" | "bool"   => BooleanType
+      case "byte" | "tinyint"   => ByteType
+      case "short" | "smallint" => ShortType
+      case "int"                => IntegerType
+      case "long" | "bigint"    => LongType
+      case "float"              => FloatType
+      case "double"             => DoubleType
+      case "string"             => StringType
+      case "binary"             => BinaryType
+      case _ => {
+        UDTCollection.get(typeName) match {
+          case Some(dt) => dt.getDeclaredConstructor().newInstance()
+          case None =>
+            throw new SchemaParseException(s"Dose not recognize ${typeName}")
+        }
+      }
+    }
+  }
+}
+
+private[sql] object SchemaParser {
+
+  private val builder = new SchemaBuilder()
+
+  /** Parse a schema simpleString with UDTs */
+  def parse(schema: String): DataType = {
+    parseSchema(schema) { parser =>
+      {
+        builder.visit(parser.schema()) match {
+          case dataType: DataType => dataType
+          case _ =>
+            throw new SchemaParseException(
+              s"does not recognize schema: ${schema}"
+            )
+        }
+      }
+    }
+  }
+
+  private def parseSchema[T](
```

Comment:
nit: what's the purpose for making this a higher-ordered function instead of just returning a parser?

Reply:
This was the pattern used in delta lake and spark 
https://github.com/delta-io/delta/blob/master/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala#L83

No sure why it does this tho.